### DISCUSSION TITLE
[QUESTION] What underlying technology is LH using to do snapshots

### DISCUSSION THREAD

**Discussion by PhilipMay** [longhorn/longhorn#3551] [Q&A]
What underlying technology is LH using to do snapshots?

Is it using LVM or some mechanism of iscsi or something "hand written by LH"?

I am asking because I plan to use LH to do backups of Postgres. This is called File System Level Backup
but it needs a "consistent snapshot" of the data directory. I would like to understand if this is what LH is doing.
Is it as consistent as a LVM snapshot is - for example?

**Comment by derekbit** [comment]
Longhorn can do crash-consistent snapshot.
You can please check the document https://longhorn.io/docs/1.2.3/concepts/#24-snapshots  for the details.

**Comment by PhilipMay** [comment]
Thanks @derekbit - if you still have the time - could you perhaps tell me how this is implemented?
Reusing somthing from iscsi, reusing something else or your own Longhorn implementation?

**Comment by PhanLe1010** [comment]
snapshot is a Longhorn implementation. Underlying each Longhorn volume is a collection of spare files. Each file is corresponding to a Longhorn snapshot. When you create a snapshot of a Longhorn volume, Longhorn does:
* running sync to flush data from the filesystem to the block device level managed by Longhorn. https://longhorn.io/docs/1.2.3/concepts/#231-how-read-and-write-operations-work-for-replicas
* Create a new sparse file for volume head. The current spare file becomes the newly created snapshot. 


@shuo-wu will share a presentation for more info

**Comment by shuo-wu** [comment]
There is a presentation. It contains some detailed illustrations and animations that may help you understand Longhorn implementations:
https://docs.google.com/presentation/d/1SAGUKE9gqWCRWtJ8li0KWqnvMmtY22LY/edit?usp=sharing&ouid=108339013651336965146&rtpof=true&sd=true
